stubs/bleach/bleach/html5lib_shim.pyi

```diff
@@ -1,27 +1,59 @@
-from _typeshed import Incomplete
-from collections.abc import Generator, Iterable
+import re
+from collections.abc import Iterable, Iterator
+from typing import Any
 
-class HTMLParser:  # actually html5lib.HTMLParser
-    def __getattr__(self, __name: str) -> Incomplete: ...
+# We don't re-export any `html5lib` types / values here, because they are not
+# really public and may change at any time. This is just a helper module,
+# import things directly from `html5lib` instead!
+from html5lib import HTMLParser
+from html5lib._inputstream import HTMLUnicodeInputStream
+from html5lib._tokenizer import HTMLTokenizer
+from html5lib.serializer import HTMLSerializer
 
-class Filter:  # actually html5lib.filters.base.Filter
-    def __getattr__(self, __name: str) -> Incomplete: ...
+HTML_TAGS: list[str]
+HTML_TAGS_BLOCK_LEVEL: frozenset[str]
+AMP_SPLIT_RE: re.Pattern[str]
+ENTITIES: dict[str, str]
+TAG_TOKEN_TYPES: set[int]
+TAG_TOKEN_TYPE_CHARACTERS: int
+TAG_TOKEN_TYPE_END: int
+TAG_TOKEN_TYPE_PARSEERROR: int
+TAG_TOKEN_TYPE_START: int
 
-class SanitizerFilter:  # actually html5lib.filters.sanitizer.Filter
-    def __getattr__(self, __name: str) -> Incomplete: ...
+class InputStreamWithMemory:
+    position: int
+    def __init__(self, inner_stream: HTMLUnicodeInputStream) -> None: ...
+    def reset(self) -> None: ...
+    @property
+    def errors(self) -> list[str]: ...
+    @property
+    def charEncoding(self) -> tuple[str, str]: ...
```

Comment:
```suggestion
    def charEncoding(self) -> tuple[str | None, str]: ...
```

`HTMLUnicodeInputStream.charEncoding` is set here; the first element of the `tuple` is whatever `lookupEncoding` returns: https://github.com/html5lib/html5lib-python/blob/f7cab6f019ce94a1ec0192b6ff29aaebaf10b50d/html5lib/_inputstream.py#L185

`lookupEncoding` can return `None`: https://github.com/html5lib/html5lib-python/blob/f7cab6f019ce94a1ec0192b6ff29aaebaf10b50d/html5lib/_inputstream.py#L903-L918

Reply:
Good point about `utf-8` being hardcoded in, but what if an `AttributeError` is raised [here](https://github.com/html5lib/html5lib-python/blob/f7cab6f019ce94a1ec0192b6ff29aaebaf10b50d/html5lib/_inputstream.py#L912-L916)? Wouldn't it still return `None` in that eventuality? (I'm not sure why an `AttributeError` _would_ be raised there, but the code seems to be specifically guarding against that happening...)

Great point about it returning an `Encoding` instance rather than a `str`!

Reply:
The only `AttributeError` I can see is if `label` doesn't have a `strip` method: https://github.com/gsnedders/python-webencodings/blob/master/webencodings/__init__.py#L74 (None or wrong object)

Reply:
Alright, sounds reasonable! Strange code at runtime, but let's assume it can't be `None` for now 👍